Add overlap with DDP to ZeRO

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62052 Add overlap with DDP to ZeRO**

Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

Commit:
Add overlap with DDP to ZeRO

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -856,6 +857,7 @@ def forward(self, *inputs, **kwargs):
             # This should be called only once during whole training period.
             if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
                 logging.info("Reducer buckets have been rebuilt in this iteration.")
+                self._has_rebuilt_buckets = True
```
Note that this might not be entirely accurate. 

If find_unused_params=True then we don't actually rebuilt buckets and it's basically a noop (you can confirm by seeing reducer code). Not sure if it really affects your use case though.

andwgu replied:
I see. The idea is that I just need a signal that buckets have been finalized since the design I preferred was to only fully initialize the `ZeroRedundancyOptimizer` instance after that point. This made sense to me given that the current DDP implementation rebuilds buckets early and never again, so any initial construction that seems wasteful, as it will soon be thrown away.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
```
This means we don't do optimizer step at all before buckets are rebuilt?

andwgu replied:
Yes. Sorry that I have not had a chance to write the PR text yet explaining the changes. This overlap implementation requires two dummy iterations to warm up, and the training loop still requires `ZeroRedundancyOptimizer.step()` to be called. It becomes a no-op once warmed up.

mrshenli replied:
(can be done in followup PRs) is it possible to let ZeRO to adapt to DDP's rebuilt buckets?

mrshenli replied:
Here, we might also need to get the Future as argument, and return `[Future.wait()]`, otherwise `wait()` on the child future might not wait for the allreduce to finish.

andwgu replied:
Commenting for completeness:

The `Future.wait()` here is not needed if using a single stream for computation since `then()` performs the needed synchronization to guarantee the all-reduce finishes. However, it is needed if the optimizer computation occurs in a separate stream. (Note: The API has changed so that the return should look like `return bucket.get_tensor()` now. However, the current all-reduce hook returns a singleton `List[Tensor]` still, so the call would look like `return fut.wait()[0]`, where `fut` is the argument to `zero_step()`.)

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+                    "Overlapping DDP with ZeRO only supports a single " \
+                    "parameter group"
+                # Construct the `gradients` input for the local optimizer step,
+                # which expects `None` in a list position to indicate that the
```
That might be a little bit brittle or easy for future developer to forget, can you do something like have a global:

_ZERO_NO_UPDATE_PARAM = None 

And use that constant instead, might make the intent more clear.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
```
I don't see a state transition to initialized, where does that happen?

andwgu replied:
It happens in [`ZeroRedundancyOptimizer.step()`](https://github.com/pytorch/pytorch/blob/f60106a46aecca09ffa75cb13a179833e6aaff2d/torch/distributed/optim/zero_redundancy_optimizer.py#L831). My understanding: Buckets are rebuilt during the second forward pass; the communication hooks collect the necessary information in the second backward pass; and the state transition happens in the second optimizer step.

rohan-varma reviewed (commented):
Just a couple comments on the hook, will take a look at Zero side changes soon.

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
```
without get the future argument and calling `Future.wait()` how do we guarantee that ZeRO computation starts after the AllReduce is done?

andwgu replied:
I believe that `wait()` is called in the `Reducer` on the future returned by the hook, according to Rohan [here](https://github.com/pytorch/pytorch/pull/61678/files#r674440741). I do not have the code pointer at the moment, but that would explain why we do not need a `wait()` call here. I am not sure if this is accurate since I think that I needed the `wait()` call in the two-stream setting to ensure correctness.

Also, for reference, [Rohan's PR for `hook_then_optim()`](https://github.com/pytorch/pytorch/blob/00f3289d7d4c4ce999a07a6f1e8630fdb1cada38/torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py#L94).

andwgu replied:
I think perhaps this is also due to the semantics of [`then()`](https://pytorch.org/docs/stable/futures.html#torch.futures.Future.then):

> In other words, as long as the callback doesn’t switch streams, it can safely manipulate the result without any additional synchronization.

mrshenli replied:
> I think perhaps this is also due to the semantics of then():

I see. Makes sense, this is enforced here:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue_inl.h#L649

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+                params_per_rank[rank_to_update].extend(bucket_params)
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
```
ditto

andwgu replied:
See https://github.com/pytorch/pytorch/pull/62052#discussion_r675670710.

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
+
+            return [bucket.get_tensor()]
```
Would I be correct if I assume that since this returned gradients won't be used anyway, we can return arbitrary Tensor values?

andwgu replied:
I do not believe that assumption is correct. I think that we must still return the bucket's tensor since the value of the new `Future` returned by `then()` will be used in `Reducer::finalize_backward()`. See [here](https://github.com/pytorch/pytorch/blob/04bd9d7577ff855419adca1eb59b58d0f73a1a18/torch/csrc/distributed/c10d/reducer.cpp#L1449).

rohan-varma replied:
Although maybe it shouldn't matter, since the returned gradients are basically just copied back into `param.grad` field, but param.grad field may be irrelevant if we are doing an explicit step on the gradients here.

Although I think you mentioned that before buckets are rebuilt user will still call `zero.step()`, in that case the correct gradient tensor indeed needs to be returned.

andwgu replied:
> Although I think you mentioned that before buckets are rebuilt user will still call zero.step(), in that case the correct gradient tensor indeed needs to be returned.

To clarify this, the `ZeroRedundancyOptimizer.step()` called before buckets are rebuilt do not actually do any optimization, so given what @rohan-varma explained, I think actually we do not need to return the bucket tensor -- just a tensor of the same dimension. I misspoke in the previous comment.

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -0,0 +1,110 @@
+            device = bucket_params[0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
```
does it make sense to hold on to the returned work handle here and in the last bucket wait for them all. So that when backward returns, all optimizer step comp/comm ops are inserted into the stream or correctly synchronized?

andwgu replied:
Good catch. I think that this is needed.

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

Commit:
Update on "Add overlap with DDP to ZeRO"




[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

Commit:
Update on "Add overlap with DDP to ZeRO"


Updated this version to only begin optimizer computation after the backward computation has completed -- hence, the goal is to overlap optimizer computation with the outstanding all-reduces for gradient synchronization.

Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

Commit:
Update on "Add overlap with DDP to ZeRO"


Updated this version to only begin optimizer computation after the backward computation has completed -- hence, the goal is to overlap optimizer computation with the outstanding all-reduces for gradient synchronization.

Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

Commit:
Update on "Add overlap with DDP to ZeRO"


Updated this version to only begin optimizer computation after the backward computation has completed -- hence, the goal is to overlap optimizer computation with the outstanding all-reduces for gradient synchronization.

Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

Commit:
Update on "Add overlap with DDP to ZeRO"


Differential Revision: [D29895933](https://our.internmc.facebook.com/intern/diff/D29895933)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29895933).

andwgu commented:
Combined the two approaches; refer to this other PR: https://github.com/pytorch/pytorch/pull/62157
